### PR TITLE
crimson/osd: fix SnapTrimEvent lifetime

### PR DESCRIPTION
**Pull request by Matan-B**
Previous attempt is reverted since the issue was still occurring.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2023-11-29_08:30:41-crimson-rados-wip-matanb-crimson-snaptrimevent-lifetime-v2-distro-crimson-smithi/

**Comment by Matan-B** [comment]
Unrelated cleanup and fixes moved to: #55101

File: src/crimson/osd/pg.cc [hunk 1]
```diff
@@ -1519,6 +1543,7 @@ void PG::on_change(ceph::os::Transaction &t) {
   recovery_backend->on_peering_interval_change(t);
   backend->on_actingset_changed(is_primary());
   wait_for_active_blocker.unblock();
+  std::ignore = snap_trim_event_gate.close();
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:1547]: I'm not sure I understand the purpose of this commit -- we aren't blocking on the gate so we don't prevent the snap trim event dispatch from continuing after the on_change call.  Moreover, I don't actually want to block on all outstanding events completing -- we generally choose instead to check for interval changes in such events using interruptible_future.

**Comment by athanatos** [comment]
All but the last two commits look reasonable to me -- did your testing suggest that reordering PG and handle wasn't sufficient?

**Comment by Matan-B** [comment]
> All but the last two commits look reasonable to me -- did your testing suggest that reordering PG and handle wasn't sufficient?

Merged in #55101, haven't seen the issue in the latest test, will update the tracker if seen again.

**Pull request closed without merging**